Add support for fixed points

randomnetcat opened the pull request:
Should be feature complete.

Tests in progress.

Fixes #409.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,54 @@ void CompilerUtils::convertType(
 			if (auto typeOnStack = dynamic_cast<IntegerType const*>(&_typeOnStack))
-				if (targetFixedPointType.numBits() > typeOnStack->numBits())
+            {
+				if (targetFixedPointType.numBits() > typeOnStack->numBits() + targetFixedPointType.fractionalDigits() * 3)
```
Can you explain the calculation here? I think it is more complicated than just multiplying the fractional digits by 3.

randomnetcat replied:
Yeah, the idea was to save a bit cleaning, but I just completely messed it up. I will remove it.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,54 @@ void CompilerUtils::convertType(
+
+            if (auto typeOnStack = dynamic_cast<RationalNumberType const*>(&_typeOnStack))
+			{
+				if (targetFixedPointType.identifier() == (*typeOnStack->fixedPointType()).identifier())
```
The target fixed point type can never be a rational number type (It is a fixed point type).

meowingtwurtle replied:
Couldn't it be a fixed point with a different number of bits or fractional digits?

randomnetcat replied:
I'm not comparing it to a rational type, I'm comparing it to the rational's minimal fixed point type.

chriseth replied:
Ah I'm sorry, I somehow misread.
Could you still do the comparison on the types themselves and not on their identifiers?

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,54 @@ void CompilerUtils::convertType(
+					m_context << typeOnStack->literalValue(nullptr);
+				} else
+				{
+					convertType(*typeOnStack, *typeOnStack->fixedPointType());
```
This looks like infinite recursion to me...

randomnetcat replied:
See above

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,54 @@ void CompilerUtils::convertType(
+				}
+			}
+
+			if (auto typeOnStack = dynamic_cast<FixedPointType const*>(&_typeOnStack)) {
```
Can you change this to a plain else plus a `solAssert`?

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,54 @@ void CompilerUtils::convertType(
+				}
+				else if (typeOnStack->fractionalDigits() < targetFixedPointType.fractionalDigits())
+				{
+					m_context << u256(pow(bigint(10), targetFixedPointType.fractionalDigits() - typeOnStack->fractionalDigits()))
```
Why don't you clean the higher order bits here? Yes, it might not be needed, but in this case, please add a comment :)

randomnetcat replied:
I'll just move the above cleaning out of the conditional.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -1046,6 +1082,11 @@ unsigned CompilerUtils::loadFromMemoryHelper(Type const& _type, bool _fromCallda
 	return numBytes;
 }
 
+IntegerType const& CompilerUtils::fixedPointTypeAsIntegerType(FixedPointType const &_type)
```
You cannot return a temporary value by reference, so this sthould be just `IntegerType`.

chriseth replied:
And yes, C++ is a really shitty language for not auto-detecting this...

randomnetcat replied:
I tried using a non reference but I got something about IntegerType not being copyable.

chriseth reviewed (changes requested):
Wonderful start! Please also correct the coding style: Indent using tabs only and put `{` on the next line, omitting them altogether if the body consists only of a single statement.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,49 @@ void CompilerUtils::convertType(
+					cleanHigherOrderBits(targetFixedPointType);
+				}
+            }
+            if (stackTypeCategory == Type::Category::RationalNumber)
```
Please use `else if` here.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -1046,6 +1077,12 @@ unsigned CompilerUtils::loadFromMemoryHelper(Type const& _type, bool _fromCallda
+shared_ptr<IntegerType> CompilerUtils::fixedPointTypeAsIntegerType(FixedPointType const &_type)
+{
+//	return IntegerType(_type.numBits(), _type.isSigned() ? IntegerType::Modifier::Signed : IntegerType::Modifier::Unsigned);
+	return make_shared<IntegerType>(_type.numBits(), _type.isSigned() ? IntegerType::Modifier::Signed : IntegerType::Modifier::Unsigned);
```
Yep, this is what I would have suggested :)
By the way, this function could be handy as part of the `FixedPointType` class.

chriseth commented:
Apart from minor details and formatting, this is ready to go :)

randomnetcat commented:
Changed the `if` to `else if`
Added `asIntegerType` to `FixedPointType`
Added support for arithmetic
Added retrieval from storage
Added ABI conversions

VoR0220 commented:
I rejoice!

meowingtwurtle commented:
DO NOT MERGE. Not tested at all.
There seems to be a bug somewhere in the compiler. Decimal constants are truncated.

VoR0220 commented:
Wow...that was stupid of C++...doesn't appear to be a bug at all.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -215,8 +215,16 @@ string ABIFunctions::cleanupFunction(Type const& _type, bool _revertOnFailure)
 			break;
 		case Type::Category::FixedPoint:
-			solUnimplemented("Fixed point types not implemented.");
+		{
```
This seems to be identical to the respective integer type. In view of the optimizer, it might be bettor to just use
`return cleanupFunction(type.integerType())`.

chriseth replied:
Or something like
```
templ("body", "cleaned := " + cleanupFunction(type.integerType() + "(value)");
```

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +317,33 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				}
+				else if (fromCategory == Type::Category::RationalNumber)
+				{
+					RationalNumberType const& fromRationalType = dynamic_cast<RationalNumberType const&>(_from);
```
Can you move this to right after the `else if (toCategory == Type::Category::FixedPoint)`?

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +317,33 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+						Whiskers("converted := <cleanFixed>(mul(<cleanInt>(value), <factor>))")
+						("cleanInt", cleanupFunction(_from, false))
+						("cleanFixed", cleanupFunction(_to, false))
+						("factor", toString(pow(bigint(10), dynamic_cast<FixedPointType const &>(_to).fractionalDigits())))
```
No space in front of `&`.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +317,33 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+					body =
+						Whiskers("converted := <cleanFixed>(mul(<cleanInt>(value), <factor>))")
+						("cleanInt", cleanupFunction(_from, false))
+						("cleanFixed", cleanupFunction(_to, false))
```
Do we really need to call `cleanFixed` in this case?

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +317,33 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				else if (fromCategory == Type::Category::RationalNumber)
+				{
+					RationalNumberType const& fromRationalType = dynamic_cast<RationalNumberType const&>(_from);
+					const shared_ptr<const FixedPointType> &fromFixedType = fromRationalType.fixedPointType();
```
`const` is always a suffix (this and the following two lines).

chriseth replied:
Also, you cannot store a reference here, it has to be a copy.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +317,33 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+						("cleanRational", cleanupFunction(_from, false))
+						("fixSize", toFractionalDigits > fromFractionalDigits ? "mul" : "div")
+						("factor", toString(pow(bigint(10), abs(toFractionalDigits - fromFractionalDigits))))
+						("cleanFixed", cleanupFunction(_to, false))
```
Also here: Probably no need to call the cleanup function.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -354,8 +386,43 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+			}
+			else if (toCategory == Type::Category::FixedBytes)
+			{
+				FixedBytesType const &toBytesType = dynamic_cast<FixedBytesType const &>(_to);
```
Perhaps it is better to just call `convertToBytes(convertToInteger(value))`.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -354,8 +386,43 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+			if (toCategory == Type::Category::Integer)
+			{
+				body =
+						Whiskers("converted := <cleanInt>(div(<cleanFixed>(value), <divisor>))")
```
This should be `sdiv` for signed.

In general, I think it would be better to add a new function that shifts by powers of ten, one for signed, one for unsigned with the amount of powers being part of the name of the function. There is already a pair of functions `shiftLeftFunction` and `shiftRightFunction` - this would basically be the same thing just for powers of ten.

randomnetcat replied:
I think that's a bit overkill.

chriseth replied:
To be honest, I don't think so. It is really easy to make a mistake here. If there are multiple places to make a mistake, it is even the harder to find. If a mistake affects all uses of the same function, it is easy to spot.

VoR0220 replied:
@meowingtwurtle overkill prevents bugs that lead to millions of dollars being lost. Just do the overkill.

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,43 @@ void CompilerUtils::convertType(
-					cleanHigherOrderBits(*typeOnStack);
-			solUnimplemented("Not yet implemented - FixedPointType.");
+			if (stackTypeCategory == Type::Category::Integer)
+            {
```
There are still spaces here :)

chriseth reviewed (changes requested):
This already looks really good, but please keep this focused on only a small aspect and don't add more features.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,13 +1484,61 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+			// stack: exp base
+
+			// Test for non-integer exponent
+			m_context << Instruction::DUP2 		// exp base exp
```
You can (and actually should) use `appendInlineAssembly()` here to increase readability by some orders of magnitude :)

randomnetcat replied:
I'll do that and convert it to use a loop to decrease overflows.

VoR0220 replied:
Why would you use a loop to decrease overflows? (I hope you didn't mean assembly loop and meant something else).

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -311,7 +308,32 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+					body =
+						Whiskers("converted := <toFixed>(<cleanInt>(value))")
+						("toFixed", scaleFixedFunction(0, toFixedPointType.fractionalDigits(),
+													   dynamic_cast<IntegerType const &>(_from).isSigned()))
```
Please format as follows:
```
x(
    a,
    b,
    c
)
```
I.e. if you want to break a line inside an argument list, put every single argument on its own line, everything that precedes the list and everything that follows the list on its own line. Indentation of the arguments is exactly one tab more than the surrounding lines.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -354,8 +376,36 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+			if (toCategory == Type::Category::Integer)
+			{
+				body =
+						Whiskers("converted := <toInt>(<cleanFixed>(value))")
```
Please only indent one additional tab.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -354,8 +376,36 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				FixedPointType const &toFixedType = dynamic_cast<FixedPointType const &>(_to);
+				body =
+						Whiskers("converted := <scale>(<cleanFixed>(value))")
+						("scale", scaleFixedFunction(fromFixedType.fractionalDigits(), toFixedType.fractionalDigits(),
```
If you break a line inside an argument list, please put all arguments on their own line (see above).

chriseth commented on libsolidity/codegen/CompilerUtils.h:
```diff
@@ -258,6 +258,11 @@ class CompilerUtils
 	void cleanHigherOrderBits(IntegerType const& _typeOnStack);
 
+	/// Appends code that cleans higher-order bits for fixed point types.
+	void cleanHigherOrderBits(FixedPointType const& _typeOnStack) {
```
Please put the `{` on its own line.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1463,6 +1471,10 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 	case Token::Mul:
 		m_context << Instruction::MUL;
+		if (fixedType)
+			m_context << u256(pow(bigint(10), fixedType->fractionalDigits()))
```
Unfortunately, it's not that simple. This works for fixed point types with up to 128 bits, but you have to deal with overflow if you have larger numbers.

chriseth replied:
I would propose to remove this feature for now, throw a "not implemented" exception for all operations where overflow might happen for fixed point operands and deal with it in a subsequent pull request.

chriseth replied:
Another thing we could do here is to change the result type of the operation, but we don't do that for integers, so it would probably be weird.

randomnetcat replied:
I think it's slightly more complicated than 128 bits. For example, a 248 bit fixed point with only 1 decimal digit could safely be shifted. I think the formula is `canScale = bits + ceil(log2(pow(10, fractionalDigits))) <= 256`

randomnetcat replied:
This also affects exponentiation and division

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,13 +1484,57 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+			std::string code = R"(
+			{
+				if lt(exponent, 0) {
+					invalid()
```
I would actually prefer to disallow that as part of the type system: `a**b` is invalid if `b` is a signed fractional type.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10556,6 +10556,393 @@ BOOST_AUTO_TEST_CASE(snark)
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(ufixed128x19)", [](u256 const &a) -> u256
+	{ return a; }, 0, 100);
```
Please correct indentation.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10556,6 +10556,393 @@ BOOST_AUTO_TEST_CASE(snark)
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed128x19)", [](u256 const &a) -> u256
+	{ return a; }, -50, 50);
```
Please correct indentation.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10556,6 +10556,393 @@ BOOST_AUTO_TEST_CASE(snark)
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(createFixed(375, 100, 19)));
```
There is a macro called `ABI_CHECK` which produces better formatted output in case of an error, please use that.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10556,6 +10556,393 @@ BOOST_AUTO_TEST_CASE(snark)
+	compileAndRun(sourceCode, 0, "test");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(createFixed(-42875, 1000, 19)));
+}
+
```
Please add tests with different fixed point types and tests that can create overflow

randomnetcat commented:
Sorry about all of the formatting things, CLion apparently has its own ideas about what is correct that I haven't gotten around to fixing

randomnetcat commented:
More tests are on the way

meowingtwurtle commented:
More tests are on the way

VoR0220 commented on test/libsolidity/SolidityExecutionFramework.h:
```diff
@@ -75,6 +75,14 @@ class SolidityExecutionFramework: public dev::test::ExecutionFramework
 
+	using rational = boost::rational<dev::bigint>;
+
+	static u256 createFixed(s256 _numerator, s256 _denominator, unsigned int _decimalDigits)
```
++

randomnetcat replied:
Um what?

VoR0220 replied:
It mean good work. 👍

randomnetcat replied:
Ah, thanks

randomnetcat commented:
Let me know when this is to be merged and I'll rebase the commits together

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -596,6 +596,12 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
 		return TypePointer();
+	if (_operator == Token::Exp)
+	{
+		if (auto intType = dynamic_pointer_cast<IntegerType const>(commonType))
```
I actually have the impression that we should not base this on the common type alone. Perhaps @axic should also give his opinion. I guess removing exp from this PR altogether might be a good thing to do.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -313,8 +313,7 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
 						Whiskers("converted := <toFixed>(<cleanInt>(value))")
-						("toFixed", scaleFixedFunction(0, toFixedPointType.fractionalDigits(),
-													   dynamic_cast<IntegerType const &>(_from).isSigned()))
+						("toFixed", scaleFixedFunction	(0, toFixedPointType.fractionalDigits(), dynamic_cast<IntegerType const &>(_from).isSigned()))
```
No space between function name and opening parethenses.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1451,6 +1451,8 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 	auto const* intType = dynamic_cast<IntegerType const*>(&_type);
 	auto const* fixedType = dynamic_cast<FixedPointType const*>(&_type);
 
+	auto const fixedFactor = fixedType ? u256(pow(u256(10), fixedType->fractionalDigits())) : u256(0);
```
From https://github.com/ethereum/solidity/pull/3389/files/089e000b86e7807bdf6dc358c303b4fe448c83aa..cfbc94dce4d087d103e62a64e62cca7ff5dfcefa#diff-07335c711d9742b9c53a75648c7ad55cR501 it could actually be that `10**fractionalDIgits` is `2**256` (not possible in this specific case, but still) - so perhaps rather use bigint for this?

Also, is the result of `pow` not a `u256` or why are you converting again?

chriseth replied:
Oh and should it not be 1 for an integer instead of 0?

randomnetcat replied:
I had just set it to zero for integers since it's not used for integers.

randomnetcat replied:
And for the overflow, that is guaranteed not to happen due to the (newly updated) check in FixedPointType's constructor.

chriseth replied:
Hm, does this zero lead to an error further down?

The reason I'm asking is that we should never ever assume any consistency between parts of the code that are more than 50 lines away from each other. Always use asserts to check any assumptions you might have about different parts of the code. This is very useful if different people work on the same code and it also helps with reviews since it catches errors that might be hidden because a reviewer did not think about parts of the code not visible in the changeset.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,9 +1474,12 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 		if (fixedType)
-			m_context << u256(pow(bigint(10), fixedType->fractionalDigits()))
+		{
+			solUnimplemented("Not yet implemented - FixedPointType multiplication");
```
Can you enable it at least for the cases where there cannot be any overflow? I.e. where
`fixedFactor < (u256(1) << (numBits / 2))` or something like that?

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1504,12 +1510,9 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
-				}
-
 				let factor := <factor>
 				if iszero(eq(exponent, mul(div(exponent, factor), factor))) { // Check if exponent has no decimal value
```
s/decimal/integer/

randomnetcat replied:
I'm sorry, what?

chriseth replied:
I think you should rather say `check if exponent has integer value` or `non-fractional` - the word `decimal` is too ambigous in this context: `10` is a decimal number (because it is written using the decimal system), but it is not fractional.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1529,8 +1532,7 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 				exponent := result // Make accessible to outside world
```
Ah, now I read the algorithm while being properly awake :)
I fear this won't do - the loop takes way too much time, we should use some square and multiply-like algorithm, but not sure if that is feasible for fixed points.

randomnetcat commented:
So 6 tests are failing, all to do with fixed point types. These seem to be failing to to the changed check in FixedPointType's constructor. This brings up the question of how the check should work.

chriseth commented:
@meowingtwurtle ok, I think now I understand the issue with the constructor / the reason why we did not have a check before: Is there a real reason why we want `1` to be always part of the range of a fixed point type? People might be ok with `fixed8x50` if they only work with very small numbers. Would that be counter-intuitive?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -596,17 +596,23 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
 	if (Token::isBitOp(_operator) || Token::isBooleanOp(_operator))
 		return TypePointer();
-	if (auto fixType = dynamic_pointer_cast<FixedPointType const>(commonType))
+	if (_operator == Token::Exp)
```
Please disallow exponentiation for now. I think it should return the base type and not the common type to be consistent with `<<`, but I'm not sure. What do you think, @axic?

axic replied:
I think it should be consistent with how it works on integer types. That works on the common type. I don't actually see a good reason right now to do exponentiation on fixed point types, so would keep it disabled, and reconsider when the need arises.

axic replied:
Also, integer types disallow exponentation with fixed point types, at the bare minimum the same restriction should apply here.

chriseth commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -354,8 +375,35 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
 		case Type::Category::FixedPoint:
-			solUnimplemented("Fixed point types not implemented.");
+		{
+			FixedPointType const &fromFixedType = dynamic_cast<FixedPointType const &>(_from);
```
The `&` is glued to the type, not the name, i.e. it should be `FixedPointType const& fromFixedType`

chriseth commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -543,17 +542,43 @@ void CompilerUtils::convertType(
-			solUnimplemented("Not yet implemented - FixedPointType.");
+			if (stackTypeCategory == Type::Category::Integer)
+			{
+				auto typeOnStack = dynamic_cast<IntegerType const*>(&_typeOnStack);
```
Please always use `dynamic_cast<IntegerType const&>(...)`, i.e. a reference if you know (or assume) that this is the correct type. The reason is that `dynamic_cast` throws an exception if the type is wrong for a reference but returns a null pointer for a pointer.

chriseth replied:
Oh and then you also have to use `auto const&` instead of `auto`, because just a plain `auto` would result in a non-reference type and thus perform a copy.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1447,11 +1447,21 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
-		solUnimplemented("Not yet implemented - FixedPointType.");
+	u256 fixedFactor = fixedType ? pow(u256(10), fixedType->fractionalDigits()) : u256(0);
+
+	bool isSigned;
```
Please always initialize local variables of elementary type (even if you assign them later - you never know how the code will change in the future) because otherwise their value is undefined.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1447,11 +1447,21 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
+
+	bool isSigned;
+
+	if (intType)
```
This could be shortened to `bool isSigned = intType ? intType->isSigned() : fixedType->isSigned();` :)

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1447,11 +1447,21 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
+	auto const* intType = dynamic_cast<IntegerType const*>(&_type);
+	auto const* fixedType = dynamic_cast<FixedPointType const*>(&_type);
 
-	if (_type.category() == Type::Category::FixedPoint)
```
Please add `solAssert(intType || fixedType, "Expected either integer or fixed point type.");`

axic replied:
This was done.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1447,11 +1447,21 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
+	if (intType)
+	{
+		isSigned = intType->isSigned();
+	} else
```
`else` should be on a line of its own.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,13 +1490,30 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 	case Token::Exp:
-		m_context << Instruction::EXP;
+		if (intType)
+		{
```
Braces around single statements should be omitted.

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,13 +1490,30 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+			m_context << Instruction::EXP;
+		}
+		else
+		{
```
Here, we have to put braces, because the windows compiler has problems with such macros.

chriseth commented on libsolidity/codegen/LValue.cpp:
```diff
@@ -181,17 +181,15 @@ void StorageItem::retrieveValue(SourceLocation const&, bool _remove) const
-			m_dataType->category() == Type::Category::Integer &&
-			dynamic_cast<IntegerType const&>(*m_dataType).isSigned()
+				(m_dataType->category() == Type::Category::Integer && dynamic_cast<IntegerType const&>(*m_dataType).isSigned())
+				||
```
`||` goes to the end of the previous line.

chriseth commented on libsolidity/codegen/LValue.cpp:
```diff
@@ -181,17 +181,15 @@ void StorageItem::retrieveValue(SourceLocation const&, bool _remove) const
-			dynamic_cast<IntegerType const&>(*m_dataType).isSigned()
+				(m_dataType->category() == Type::Category::Integer && dynamic_cast<IntegerType const&>(*m_dataType).isSigned())
+				||
+				(m_dataType->category() == Type::Category::FixedPoint && dynamic_cast<FixedPointType const&>(*m_dataType).isSigned())
```
Perhaps it makes sense to introduce a common base class for integer and fixed point types that has an `isSigned` function...

chriseth commented:
Could you please also add some tests that cause overflow (or are close to overflow)?

Also, code that is not tested is the code that performs explicit type conversion between different fixed point types and also integers and fixed point types. Overflow tests there would also be nice.

chriseth commented:
This is almost done apart from the tests for overflow.

randomnetcat commented:
Added two basic overflow tests, some conversion tests

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,501 @@ BOOST_AUTO_TEST_CASE(snark)
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(ufixed256x19)", [](u256 const &a) -> u256 { return 1 + 0 * a; }, pow(u256(10), 19), pow(u256(10), 19) + 10);
```
Could you test the range `10**19 - 10` until `10**10 + 10`? Only that way, we find out whether the implementation rounds or truncates.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,501 @@ BOOST_AUTO_TEST_CASE(snark)
+	}
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed256x19)", [](u256 const &a) -> u256 { return createFixed(-1) + 0 * a; }, createFixed(-1, 1, 19) - 10, createFixed(-1, 1, 19));
```
Same here, plus tests for negative and positive numbers.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,501 @@ BOOST_AUTO_TEST_CASE(snark)
+	contract test {
+		ufixed256x19 a = (2**256-1)/(10**19);
+		function f() returns (ufixed256x19) {
+			return a + 1/(10**19);
```
You can also write `1e-19`

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,507 @@ BOOST_AUTO_TEST_CASE(snark)
+{
+	char const* sourceCode = R"(
+	contract test {
+		fixed256x19 a = fixed256x19(ufixed256x19((2**255-1)/(10**19))); // Compiler rejected just the plain exponential expression, though it should be within signed values. Maybe something to look in to.
```
Please remove this comment and create an issue instead :)

randomnetcat replied:
Created as #3488

chriseth commented:
OK, sorry to bother you again but please also add tests for:
 - signed overflow
 - overflow of types which do not utilize all the 256 bits, especially signed
 - explicit conversion from fixed point to int (sorry if you already have that, but did not find them with a quick glance)

chriseth commented:
@axic do we also require documentation here?

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1463,6 +1467,14 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+			solUnimplementedAssert(fixedFactor < (u256(1) << (fixedType->numBits() / 2)), "Not yet implemented - large FixedPointType multiplication");
+			m_context << fixedFactor
+					  << Instruction::SWAP1
+					  << (isSigned ? Instruction::SDIV : Instruction::DIV);
```
If you use sdiv, you have to ensure that the operands are properly cleaned before you use it. Is that the case?

chriseth replied:
I guess yse, because the sdiv has been here before - just checking. Also a test for exactly that would be nice :)

meowingtwurtle replied:
No, that is not guaranteed. The only way I could think of to ensure this would be to do a proper conversion through CompilerUtils. This probably also happens around here with multiplication.

randomnetcat replied:
Okay, will do

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1463,6 +1467,14 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 		m_context << Instruction::MUL;
+		if (fixedType)
+		{
+			// TODO: not really tested
```
Please change that :)

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1472,13 +1484,28 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+				// TODO: not really tested
+				solUnimplementedAssert(fixedFactor < (u256(1) << (fixedType->numBits() / 2)), "Not yet implemented - large FixedPointType division");
+				// Division would cancel out factor, this gives the numerator an extra factor so it remains in result
+				m_context << fixedFactor
```
Do we have tests for this at all?

randomnetcat replied:
Yes, there are tests for division, although there aren't any for divisor magnitude less than one. I'll add those.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -596,17 +595,24 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
+//		else return dynamic_pointer_cast<FixedPointType const>(commonType)->isSigned() ? TypePointer() : commonType;
 	}
 	else if (auto intType = dynamic_pointer_cast<IntegerType const>(commonType))
 		if (intType->isAddress())
```
This does not make sense here, the common type cannot be an address.

randomnetcat replied:
Yep, will fix.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,535 @@ BOOST_AUTO_TEST_CASE(snark)
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(createFixed(-225, 100, 19)));
+}
+
+BOOST_AUTO_TEST_CASE(ufixed_constant_multiplication)
```
This test actually does have no relation to fixed points, it is about rationals.

randomnetcat replied:
Yeah, I realized that. The same goes for all of the other constant arithmetic tests. Should I remove them or just rename them?

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,535 @@ BOOST_AUTO_TEST_CASE(snark)
+{
+	char const* sourceCode = R"(
+	contract test {
+		fixed a = 3.5;
```
Before we merge this, I think we should make `ufixed` be a `ufixed128x18`, so that `1 ether` can be stored as `1` in a `ufixed`

randomnetcat replied:
I can make that change, I just couldn't find where the default is in the code.

chriseth replied:
Types.cpp:200

meowingtwurtle replied:
Nevermind, found where the default is.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -10589,6 +10589,535 @@ BOOST_AUTO_TEST_CASE(snark)
+	ABI_CHECK(callContractFunction("f()"), encodeArgs(pow(u256(2), 255)));
+}
+
+BOOST_AUTO_TEST_CASE(short_ufixed_overflow)
```
Can you also add a test for multiplication overflow and perhaps also division?

I guess we even support `%` which we either shouldn't or should be clear on its semantics.

randomnetcat replied:
I'll add those tests.

Modulo is well defined on non-integer numbers. It's defined in the same way as integer modulo. Equivalently, you just multiply both operands up so that they are integers, then divide the result by whatever you multiplied by in the beginning. There is done by the fixed point factor.

chriseth replied:
ok, thanks! Then please add a test :)

axic commented:
> @axic do we also require documentation here?

Would be nice to some extent - at least the "Fixed Point Numbers" section in the types chapter.

axic commented:
Please rebase now that tests can be successfully run again 😉

chriseth commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1487,6 +1491,13 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
 		break;
 	case Token::Mul:
 		m_context << Instruction::MUL;
```
I'm not 100% sure this works correctly for signed 128 bit numbers. Could you add a test for that, please? We especially need to consider cases where the number has some dirty higher order bits. Perhaps we should clean the numbers just in case.

chriseth replied:
Added the cleanup.

chriseth commented:
@axic we should probably discuss this PR in our next meeting.

fulldecent commented:
Do we have an issue for this big change?

And is there a simple example contract that illustrates the new feature this provides?

randomnetcat commented:
This was originally kept track of in #409, which I will update if this gets merged. 

This is where most of the recent discussion has been, as well as in the Gitter.

Features are as in the documentation, although there is no specific example contract as of now.

chriseth commented:
We should not close #409 or at least create new issues for the features that are missing.

randomnetcat commented:
Got it. I'll remove the closing tag.

VoR0220 commented:
Time to redo the ERC20 standard for fixed points.

chriseth commented:
Moved to 0.5.0

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -1431,6 +1479,23 @@ string ABIFunctions::shiftRightFunction(size_t _numBits, bool _signed)
+std::string ABIFunctions::scaleFixedFunction(int fromFractionalDigits, int toFractionalDigits, bool _signed)
+{
+	string functionName = "scale_fixed_" + toString(fromFractionalDigits) + "_to_" + toString(toFractionalDigits) + (_signed ? "_signed" : "_unsigned");
+		return createFunction(functionName, [&]() {
```
Indentation seems to be off here?

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -1431,6 +1479,23 @@ string ABIFunctions::shiftRightFunction(size_t _numBits, bool _signed)
 
+std::string ABIFunctions::scaleFixedFunction(int fromFractionalDigits, int toFractionalDigits, bool _signed)
+{
+	string functionName = "scale_fixed_" + toString(fromFractionalDigits) + "_to_" + toString(toFractionalDigits) + (_signed ? "_signed" : "_unsigned");
```
Where is `toString` defined? Should've we use C++'s `to_string`?

randomnetcat replied:
For those, since they are just `int`s, yes, we can use `to_string`. Slightly below, we have to use `toString` to get the `bigint` into a string.

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -996,7 +999,7 @@ u256 RationalNumberType::literalValue(Literal const*) const
 		solAssert(fixed, "");
 		int fractionalDigits = fixed->fractionalDigits();
-		shiftedValue = (m_value.numerator() / m_value.denominator()) * pow(bigint(10), fractionalDigits);
+		shiftedValue = m_value.numerator() * pow(bigint(10), fractionalDigits) / m_value.denominator();
```
I think we should make a decision on rounding rules here, enforce and document that.

randomnetcat replied:
It should be as exact as possible, can be converted later if necessary. We shouldn't arbitrarily use information from the rational here.

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -607,17 +606,21 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
 		return commonType;
 	if (Token::isBitOp(_operator) || Token::isBooleanOp(_operator))
```
I'd just move the `== Exp` check to here.

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -607,17 +606,21 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
+//		else return dynamic_pointer_cast<FixedPointType const>(commonType)->isSigned() ? TypePointer() : commonType;
 	}
-	else if (auto intType = dynamic_pointer_cast<IntegerType const>(commonType))
-		if (intType->isAddress())
```
Actually I think there should be a check for `!isAddress` up in the main one in the beginning of the function.

axic reviewed (changes requested):
Can you please pull out the change from 19 to 18 digits and the documentation change into a new PR (and perhaps some of the operator restrictions), that I think we could merge right away.

randomnetcat commented:
Rebased

randomnetcat commented:
Documentation and 18 decimal place default added to #3801.

fulldecent commented:
I do not support the addition of fixed points to Solidity. I think this is a presentation-layer issue. Just like `printf()` formatting is not in scope for Solidity, I think non-integer types are out of scope.

I am confident that this feature will make contract auditing more difficult and will lead to lost money in the real world.

If people want to use a convention of 10^18 is one unit then they should use the `finney` and `ether` keywords which work very well and are well understood.

VoR0220 commented:
@fulldecent if it's a presentation layer issue, why is safeMath so necessary to denote decimal places in an ERC20 exchange? This isn't for the purpose of ether necessarily as it's moreso for the application layer itself ala trade.

fulldecent commented:
SafeMath addresses only integer overflows. Overflows are not presentation layer.

I am not aware of any feature of SafeMath that relates to the number of decimal places.

VoR0220 commented:
That's kind of my point. Overflows are addressed here inherently based on the types. This helps not only with presentation but also just logically speaking if we were going to be doing percentage mathematics. The feature makes the job easier, not harder, to audit would be my argument back to you. In addition the extra mathematical properties it provides enables greater financial applications would be another argument I would extend to you.

randomnetcat commented:
Also, every language has features that you don't technically need to use. Think about arrays. Technically, at least for storage arrays, you could just use a storage pointer in solidity. In some other languages, you could just use pointers to iterate through memory addresses.

axic commented:
> If people want to use a convention of 10^18 is one unit then they should use the finney and ether keywords which work very well and are well understood.

Fixed point types are not limited to 18 decimal places, that is the default setting though, similarly how `int` and `uint` are aliases to `int256` and `uint256`, `fixed` and `ufixed` are aliases for `fixed128x18` and `ufixed128x18`.

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -536,8 +536,7 @@ bool FixedPointType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 {
 	return _convertTo.category() == category() ||
-		_convertTo.category() == Category::Integer ||
-		_convertTo.category() == Category::FixedBytes;
```
I think all the changes done to `ast/Types.cpp` could be pulled into a separate PR and comprehensive test cases can be added to the `syntaxTests` directory.

As a start we need tests for:
- explicit conversions:  integers (+addresses), fixedbytes <-> fixedpoint
- binary operators: comparison, exp, bit, etc.

randomnetcat replied:
So:
 - Conversions permitted integer <-> fixed and fixed <-> fixed.
   - Addresses fail
 - Comparisons.... compare (mind blown)
 - Exponentiation is not yet implemented
 - Bitwise operators fail
 - All other operators fail

Any issues with that?

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -536,8 +536,7 @@ bool FixedPointType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 	return _convertTo.category() == category() ||
-		_convertTo.category() == Category::Integer ||
-		_convertTo.category() == Category::FixedBytes;
+		_convertTo.category() == Category::Integer;
```
Hmm, should also have another check that it is not an address.

randomnetcat replied:
Done.

chriseth commented:
@fulldecent can you explain your point in more detail? Is it just a generic "as few features as possible" reason or do you have specific concerns?

The lack of fixed points is generally seen as a missing feature. My argument was always that fixed points can be added using library functions, and there actually is a library: https://github.com/NovakDistributed/macroverse/blob/master/contracts/RealMath.sol

The other argument is that `x * y` is just easier to read than `Lib.mul(x1, x2, y1, y2)`.

fulldecent commented:
My complaints are general and are specifically:

* The documentation is unclear.
* There are no examples.
* There are no test cases.
* Multiplication is undefined.
* When multiplication is defined it will need to deal with rounding.
* When multiplication is defined this equation will not hold `X * 10 * 0.1 == X`.
* There are other ways do to this already, the demand is not well motivated, and the solutions that users will choose depends on their application. Prove me wrong, link to a list of mainnet contracts that are using RealMath!

I spoke at a conference last weekend and beginners are already confused enough with reentrancy. I asked my co-panelists "what is 200 + 200 in Solidity" and they answered 400 (correct answer is 144). Before we introduce yet another caveat on common sense (X * 10 * 0.1 != X) let's carefully evaluate the demand for standardization of this approach and whether we are painting ourselves into a corner.

VoR0220 commented:
I can tell you personally that the demand has been there from within and from outside the community for years. This is partly my fault for not getting it done at an earlier date and then falling behind with other priorities. I think it would be a shame for @meowingtwurtle 's efforts to be shot down so easily when we are so close to getting this highly demanded feature over the line. 

Regarding your other points. IIRC isn't rounding already a solved problem when it comes to multiplication of fixed points? I know we can work on scaling and precision but that's for numbers at the tail end that would round off (and I'm not certain we should be hyper concerned about that in a 32 byte word system...but the edge cases are indeed worrisome...still this is basic computer science we are talking about here). 

Easy fix for the 0.1 problem is to just throw a warning whenever the constant between 0 and 1 is utilized. 

(also, your 200 + 200 claim is wrong....I even went and double checked it because I was wtf-ing after I read that)

fulldecent commented:
The demonstration of demand (citing deployed contracts on mainnet) should be well documented as motivation for changing the language.

I have not shot down this work. I have raised objections which are overcomeable. Some are difficult to overcome.

Multiplication is not solved. This is why we do not have floating point numbers in Solidity, among other reasons.

Warning on numbers between 0 and 1 is wow.

If we need to make a decision about rounding for multiplication then this should be the highlight of our discussion here. Including examples of contracts deployed on mainnet that are using Mul(a,b,c,d) functions as above. And also including a discussion of other use cases which might have a problem with the proposed solution.

The time to consider changes to a language is after a careful analysis of the current and proposed usage is considered -- not after some code has been written.

```solidity
pragma solidity 0.4.21;

contract Add {
    function test() external pure returns (int total) {
        var a = 200;
        var b = 200;
        return a+b; // returns 144
    }

    function test2() external pure returns (int total) {
        var a = 300;
        var b = 300;
        return a+b; // returns 600
    }
}```

axic commented:
@fulldecent that is a very specific example. Note, that behaviour is documented (albeit not prominently): http://solidity.readthedocs.io/en/develop/types.html?highlight=var#type-deduction

Also note, that `var` is deprecated since 0.4.20 and will be fully removed in 0.5.0.

chriseth commented:
@fulldecent your arguments suggest that people are not used to fixed precision arithmetics, which I would doubt. We could solve the `x * 10 * 0.1 == x` issue by extending the type with each operation, but then people still have to deal with trimming the type as soon as it gets larger than a word. After all, this is a programming language and not mathematics (and even mathematics has its pitfalls with notation).

fulldecent commented:
Thank you for the discussion. I have no new arguments to add. But I will be happy to review when this is ready to merge.

I expect documentation, test cases and examples will be complete before this is merged.

randomnetcat commented:
The changes are rebasing and adding axic's check for address in explicit conversions.
The CircleCI tests failed because it apparently couldn't get the commit.

chriseth commented:
@fulldecent what other test cases do you need besides those in https://github.com/ethereum/solidity/pull/3389/files#diff-4d1a8dae4b37bd7d3d29302cf7345481R11144 ?

randomnetcat commented:
@chriseth @fulldecent I did some testing. It appears that, because of the encoding that is used (stored as multipled by 10 to the power of <decimals>), that your equality will hold for certain numbers. `x * (1/a) * a == x` will hold for all numbers a such that both `a` and `1/a` are terminating decimals in base 10 (size constraints not withstanding, so the 18th decimal place would be lost for `ufixed128x18`).

Yes, this includes `0.1 * x * 10 == x` and even `32/5000 * x * 5000/32 == x`.
However, this does mean that `(1/3) * x * 3 != x`.

This is all fractions in simplest form such that both the numerator and denominator prime factorize into only 2's and 5's.

randomnetcat commented:
Full documentation is also coming.

chriseth commented:
@meowingtwurtle please check that your precedence is right, `(.1 * x) * 10` should always lose precision, because it rounds after the first multiplication.

randomnetcat commented:
Yes, it does lose precision, but it is well-defined. And, unless there is a value in the last decimal slot, it does not actually lose precision.

fulldecent commented:
@chriseth the notable test case 1/3 is missing.

randomnetcat commented:
@chriseth @fulldecent Test case added, documentation updated, and rebased. I also changed the check for large fixed point multiplication/division in 39e4dca. Could someone with slightly more knowledge of how arithmetic on EVM works than me please review that?

axic commented:
Can you squash some commits and rebase over #3958 ?

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -1434,6 +1482,23 @@ string ABIFunctions::shiftRightFunction(size_t _numBits, bool _signed)
 	});
 }
 
+std::string ABIFunctions::scaleFixedFunction(int fromFractionalDigits, int toFractionalDigits, bool _signed)
```
Should these parameters be signed?

randomnetcat replied:
No, there cannot be negative fractional digits.

axic replied:
I think it probably makes sense changing that, but would do that in a separate PR, because otherwise it may be too many conflicting changes again.

randomnetcat replied:
Oh, it might mess up the subtraction down below, but it shouldn't really matter since the range is fine either way.

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1645,13 +1656,27 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+			m_context << Instruction::EXP;
+		else
+		{
+			solUnimplemented("Not yet implemented - FixedPointType exponentiation");
```
Should this be a regular assert instead because it is disabled in the type system? Also probably swap the if statement to match the others: `if (fixedType)`

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1645,13 +1656,27 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		{
+			if (fixedType)
+			{
+				solUnimplementedAssert(fixedFactor < pow(bigint(2), 256 - fixedType->numBits()), "Not yet implemented - large FixedPointType division");
```
This seems to be duplicated 3 times, can this be improved?

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -1434,6 +1482,23 @@ string ABIFunctions::shiftRightFunction(size_t _numBits, bool _signed)
+			)")
+			("functionName", functionName)
+			("scaleFunc", toFractionalDigits >= fromFractionalDigits ? "mul" : _signed ? "sdiv" : "div")
+			("factor", toString(pow(bigint(10), abs(fromFractionalDigits - toFractionalDigits))))
```
Shouldn't there be an assertion for the upper bound of this factor?

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -314,7 +311,31 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				{
+					RationalNumberType const& fromRationalType = dynamic_cast<RationalNumberType const&>(_from);
+					shared_ptr<FixedPointType const> fromFixedType = fromRationalType.fixedPointType();
+					const int fromFractionalDigits = fromFixedType->fractionalDigits();
```
Should be an assertion here that `fromFixedType` was set properly.

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -357,8 +378,35 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+			}
+			else if (toCategory == Type::Category::FixedPoint)
+			{
+				FixedPointType const &toFixedType = dynamic_cast<FixedPointType const &>(_to);
```
Style: ampersand should be next to `const`

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -357,8 +378,35 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				FixedPointType const &toFixedType = dynamic_cast<FixedPointType const &>(_to);
+				body =
+					Whiskers("converted := <scale>(<cleanFixed>(value))")
+					("scale", scaleFixedFunction(fromFixedType.fractionalDigits(), toFixedType.fractionalDigits(), fromFixedType.isSigned()))
```
What if one type is signed and the other isn't?

randomnetcat replied:
We only care about if the from type is signed, since that is what's going to affect division/multiplication. Once that's done, if the other type is unsigned, it's reinterpreted as unsigned.

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -357,8 +378,35 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+					("cleanFixed", cleanupFunction(_from, false))
+					.render();
+			}
+			else if (toCategory == Type::Category::FixedBytes)
```
I think this can be removed now since it is disallowed in the type system.

randomnetcat commented:
@axic Made requested changes, although it occurred to me that there are no tests for shortening fixed point types. I'll add those now.

randomnetcat committed:
Add support for mathematical operations for FixedPointType

randomnetcat committed:
Add CompilerUtils cases for FixedPointType

randomnetcat committed:
Add LValue cases for FixedPointType

randomnetcat committed:
Add ABIFunctions cases for FixedPointType

randomnetcat committed:
Update documentation for fixed points

randomnetcat committed:
Add tests for FixedPointType

randomnetcat committed:
Require cleanup for fixed point multiplication, division and modulo.

randomnetcat committed:
Update large fixed point check for multiplication and division

randomnetcat committed:
Add fixed point shortening tests
Also fixed a style error

axic commented:
Rebased and pushed some lint changes.

@meowingtwurtle are you still working on some tests?

randomnetcat commented:
I've added all tests that I can think of. I'm of course willing to add more if there are any you think are missing.

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -314,7 +311,32 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+					body =
+						Whiskers("converted := <scale>(value)")
+						("scale",
+						 scaleFixedFunction(fromFractionalDigits, toFractionalDigits, fromRationalType.isNegative()))
```
Lint: should be on the previous line

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -314,7 +311,32 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				{
+					body =
+						Whiskers("converted := <toFixed>(<cleanInt>(value))")
+						("toFixed", scaleFixedFunction(0, toFixedPointType.fractionalDigits(), dynamic_cast<IntegerType const &>(_from).isSigned()))
```
Lint: should be nice space between `const & `

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -314,7 +311,32 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				FixedPointType const& toFixedPointType = dynamic_cast<FixedPointType const&>(_to);
+				if (fromCategory == Type::Category::Integer)
+				{
+					body =
```
I wonder if we should add an assertion here for the type width being appropriate?

randomnetcat replied:
How could a width be not appropriate?

axic replied:
Explicit conversions don't seem to check whether the target is of equal or exceeding width.

Github is sabotaging my review as it is constantly timing out :(

axic commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -314,7 +311,32 @@ string ABIFunctions::conversionFunction(Type const& _from, Type const& _to)
+				{
+					RationalNumberType const& fromRationalType = dynamic_cast<RationalNumberType const&>(_from);
+					shared_ptr<FixedPointType const> fromFixedType = fromRationalType.fixedPointType();
+					solAssert(fromFixedType, "");
```
Same here.

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,43 @@ void CompilerUtils::convertType(
+						  << Instruction::MUL;
+				if (_cleanupNeeded) {
+					cleanHigherOrderBits(targetFixedPointType);
+				}
```
Lint: these parentheses are not needed

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,43 @@ void CompilerUtils::convertType(
+				IntegerType const& typeOnStack = dynamic_cast<IntegerType const&>(_typeOnStack);
+				cleanHigherOrderBits(typeOnStack);
+				m_context << u256(pow(bigint(10), targetFixedPointType.fractionalDigits()))
+						  << Instruction::MUL;
```
Line: either place this on the same line or have a new `m_context << `here or move both onto a separate line from `m_contaxt`.

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,43 @@ void CompilerUtils::convertType(
+			else if (stackTypeCategory == Type::Category::RationalNumber)
+			{
+				auto typeOnStack = dynamic_cast<RationalNumberType const*>(&_typeOnStack);
+				convertType(*typeOnStack->fixedPointType(), targetFixedPointType);
```
Need an assert here for `typeOnStack`.

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,43 @@ void CompilerUtils::convertType(
+			{
+				solAssert(stackTypeCategory == Type::Category::FixedPoint, "");
+				auto typeOnStack = dynamic_cast<FixedPointType const*>(&_typeOnStack);
+				cleanHigherOrderBits(*typeOnStack);
```
Need an assert here for `typeOnStack`.

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,43 @@ void CompilerUtils::convertType(
+							  << Instruction::SWAP1
+							  << (typeOnStack->isSigned() ? Instruction::SDIV : Instruction::DIV);
+				else if (typeOnStack->fractionalDigits() < targetFixedPointType.fractionalDigits())
+					m_context << u256(pow(bigint(10), targetFixedPointType.fractionalDigits() - typeOnStack->fractionalDigits()))
```
I liked your scaling helper in `ABIFunctions`. Do you think a helper could be created for this file?

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1620,11 +1620,16 @@ void ExpressionCompiler::appendOrdinaryBinaryOperatorCode(Token::Value _operator
 {
-	IntegerType const& type = dynamic_cast<IntegerType const&>(_type);
-	bool const c_isSigned = type.isSigned();
+//	IntegerType const& type = dynamic_cast<IntegerType const&>(_type);
```
Is this still needed?

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -1645,13 +1657,27 @@ void ExpressionCompiler::appendArithmeticOperatorCode(Token::Value _operator, Ty
+		if (fixedType)
+		{
+			solAssert(false, "Not yet implemented - FixedPointType exponentiation");
+		}
```
Lint: unneeded parentheses

axic reviewed (changes requested):
Can you also add syntax tests?

axic commented:
Also if #4036 is merged, that should simplify a couple of things here.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	bytes const &result = callContractFunction("f()");
+	ABI_CHECK(result, encodeArgs(createFixed(-314, 100, 18)));
```
I think the variable `result` isn't really needed here.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+	)";
+	compileAndRun(sourceCode, 0, "test");
+	testContractAgainstCppOnRange("f(fixed128x18)", [](u256 const& a) -> u256 { return a; }, -50, -1);
+	testContractAgainstCppOnRange("f(fixed128x18)", [](u256 const& a) -> u256 { return a; }, 0, 50);
```
These aren't encoding the number as `fixed128x18`, but rather as `fixed256x0` strictly speaking.

axic commented on test/libsolidity/SolidityExecutionFramework.h:
```diff
@@ -87,6 +87,14 @@ class SolidityExecutionFramework: public dev::test::ExecutionFramework
 
+	using rational = boost::rational<dev::bigint>;
+
+	static u256 createFixed(s256 _numerator, s256 _denominator = 1, unsigned int _decimalDigits = 0)
```
Should this have a `totalBits` parameter (set to 256 by default)?

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+			fixed b = -2.5;
+
+			function f() returns (fixed) {
+				return a % b;
```
Why should this support negative modulo?

randomnetcat replied:
Negative modulus is well-defined.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+}
+
+/*
+BOOST_AUTO_TEST_CASE(ufixed_exponentiation)
```
Can you remove this?

randomnetcat replied:
Sure

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+			ufixed b = 1.25;
+
+			function f() returns (bool) {
+				return a > b;
```
Should also have a test for `a < b` == false.

axic commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -11901,6 +11901,566 @@ BOOST_AUTO_TEST_CASE(senders_balance)
+			fixed b = -1.25;
+
+			function f() returns (bool) {
+				return a > b;
```
Same here.

axic reviewed (changes requested):
Needs tests for `>=` and `<=`.

randomnetcat committed:
Update CompilerUtils

randomnetcat committed:
Update ABIFunctions

randomnetcat committed:
Update ExpressionCompiler

randomnetcat committed:
Update tests

axic commented on test/libsolidity/SolidityExecutionFramework.h:
```diff
@@ -89,11 +89,14 @@ class SolidityExecutionFramework: public dev::test::ExecutionFramework
-	static u256 createFixed(s256 _numerator, s256 _denominator = 1, unsigned int _decimalDigits = 0)
+	static u256 createFixed(s256 _numerator, s256 _denominator = 1, unsigned int _decimalDigits = 0, unsigned int _totalBits = 256)
 	{
+		solAssert(_totalBits <= 256, "Illegal _totalBits in createFixed");
```
I think we're not using `solAssert`s in the test suite because it throws a solidity specific exception. This one should use boost exceptions, e.g. `BOOST_REQUIRE_MESSAGE(_totalBits <= 256, ...)`

randomnetcat replied:
Okay, will update!

randomnetcat committed:
Add syntaxTests

axic commented:
Question: should we have implicit conversion between `intM` and `fixedMx0` ?

randomnetcat commented:
If `fixedMx0` is a permissible type, sure. But, I don't see why it should be. It is functionally equivalent to a `intM`.

axic commented:
> If `fixedMx0` is a permissible type, sure.

It is. Please add a test for it too :)

I'm not sure if it makes sense, but it may... though it will require a lot more tests because then in many cases a common type can be derived when doing binary operations with fixed point + integer types.

axic commented:
Can you also please rebase (again), sorry :(

randomnetcat commented:
[Following is my opinion, if you want something else, that's fine]
I think that if a developer elects to use a `fixedMx0`, that they have forfeited use of bitwise operators. `fixedMx0` shouldn't be treated any differently than other fixed point types.

axic commented:
> I think that if a developer elects to use a fixedMx0, that they have forfeited use of bitwise operators.

Sorry, I meant any arithmetic opration taking two operands, such as addition of `fixedMx0` and `intM` would find a common type and would be allowed. Because of that, probably is a bad idea (at least initially) to allow implicit conversion between `fixedMx0` and `intM`, even though there's no information loss.

The question then is: do we need `fixedMx0`?

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -2090,7 +2090,9 @@ bool ExpressionCompiler::cleanupNeededForOp(Type::Category _type, Token::Value _
 	if (Token::isCompareOp(_op) || Token::isShiftOp(_op))
 		return true;
-	else if (_type == Type::Category::Integer && (_op == Token::Div || _op == Token::Mod))
+	else if (_op == Token::Div || _op == Token::Mod)
```
Should this keep `_type == Type::Category::Integer`?

randomnetcat replied:
@chriseth (I didn't write this code)
I guess I accidentally reset author (there were some issues with timestamp and github log)

axic commented on libsolidity/codegen/ExpressionCompiler.cpp:
```diff
@@ -2090,7 +2090,9 @@ bool ExpressionCompiler::cleanupNeededForOp(Type::Category _type, Token::Value _
-	else if (_type == Type::Category::Integer && (_op == Token::Div || _op == Token::Mod))
+	else if (_op == Token::Div || _op == Token::Mod)
+		return true;
+	else if (_type == Type::Category::FixedPoint && (_op == Token::Div || _op == Token::Mul || _op == Token::Mod))
```
If the above is a generic `if div or mod` then those can be removed from here.

randomnetcat replied:
@chriseth (I didn't write this code)

axic commented:
> We should not close #409 or at least create new issues for the features that are missing.

@chriseth why is that? What is missing in this PR?

Also how should we proceed with this. I think we're slowly getting closer.

randomnetcat commented:
I think there are two non-awful options here regarding `fixedMx0`:
1) completely disallow it **[what I prefer]**
2) completely treat it as an alias to `intM` (at the parser level) _maybe unnecessarily complicated_

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -1298,5 +1325,18 @@ unsigned CompilerUtils::prepareMemoryStore(Type const& _type, bool _padToWords)
 
+void CompilerUtils::convertFixedPointType(unsigned int _from, unsigned int _to, bool _signed)
+{
+	if (_from > _to)
```
Should this have a case from `_from == _to`?

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -741,17 +740,37 @@ void CompilerUtils::convertType(
+				auto typeOnStack = dynamic_cast<FixedPointType const*>(&_typeOnStack);
+				solAssert(typeOnStack, "");
+				cleanHigherOrderBits(*typeOnStack);
+				convertFixedPointType(typeOnStack->fractionalDigits(), targetFixedPointType.fractionalDigits(), typeOnStack->isSigned());
```
Should this case have an assertion for `targetFIxedPoinType.numBits() >= typeOnStack->numBits()` ?

meowingtwurtle replied:
No, that is not a requirement. The helper function handles that.

randomnetcat replied:
Addressed in Gitter. There's a cleanup before and after, so it doesn't matter.

randomnetcat committed:
Use helper method in CompilerUtils in one more place

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -1298,5 +1323,18 @@ unsigned CompilerUtils::prepareMemoryStore(Type const& _type, bool _padToWords)
 
+void CompilerUtils::convertFixedPointType(unsigned int _from, unsigned int _to, bool _signed)
+{
+	if (_from > _to)
```
Oh, this could have an assert for `_from <= 78` and `_to <= 78` (or whatever is the limit).

Otherwise truncating the result of `pow` to `u256` could result in 0.

randomnetcat replied:
I feel like that violates separation of concerns.

axic replied:
There's no guarantee this function gets a value from a range checked previously.

randomnetcat replied:
If #4061 is addressed, it is guaranteed, since FixedPointType would be guaranteed to only have valid values and this can only be called with values from FixedPointType (or 0).

axic commented:
@ekpyron do you want to have a look at this since you've seem to have read on the fixed point discussions already ?

ekpyron commented:
@axic OK - I'll try to have a look into it one of these days!

axic commented:
Rebased, lets see.

axic commented:
I think this shouldn't be merged before #4064

randomnetcat commented:
That seems reasonable.

ekpyron commented on libsolidity/codegen/ABIFunctions.cpp:
```diff
@@ -215,8 +215,7 @@ string ABIFunctions::cleanupFunction(Type const& _type, bool _revertOnFailure)
 		case Type::Category::FixedPoint:
-			solUnimplemented("Fixed point types not implemented.");
-			break;
+			return cleanupFunction(*dynamic_cast<FixedPointType const&>(_type).asIntegerType());
```
What about the ``_revertOnFailure`` flag? Shouldn't it be passed down here?

randomnetcat replied:
Yes, yes it should. Good catch.

randomnetcat committed:
Pass _revertOnFailure for fixed point cleanupFunction in ABIFunctions

randomnetcat committed:
Update tests for 0.5.0

axic commented:
Who wants to endure the challenge of rebasing this? :)

randomnetcat commented:
Rebased.
One tiny semantic change: require cleanup for fixed points if op is Exp, even though Exp is not yet permitted, just so we don't forget in the future. Let's see if the tests pass!

0yi0 commented:
Hi, how is this PR going? I see that it was rebased but not merged.

chriseth commented:
No progress, mainly but not only because of the discussion happening here: https://github.com/ethereum/solidity/issues/409

This will probably only be implemented with the new yul-based code generator.

chriseth commented:
Closing since we are moving on to yul.
